File: samples/monorepo/ts_lib/src/exec.ts
// Intentionally insecure sample code -- Waypoint test fixture. Do not deploy.
import { exec } from "child_process";

export function evalExpr(userStr: string): unknown {
  // WAYPOINT-PLANT: TS-EVAL [axes: security]
  return eval(userStr);
}

export function makeFn(userStr: string): Function {
  // WAYPOINT-PLANT: TS-EVAL [axes: security]
  return new Function("x", userStr);
}

export function runCmd(userInput: string): void {
  // WAYPOINT-PLANT: TS-CHILD-PROCESS [axes: security]
  exec(`git log ${userInput}`, (err, stdout) => {
    if (err) return;
    console.log(stdout);
  });
}

File: samples/monorepo/rust_service/src/exec.rs
// Intentionally insecure sample code -- Waypoint test fixture. Do not deploy.
use std::process::Command;

pub fn run_user_cmd(user_input: &str) -> String {
    // WAYPOINT-PLANT: RUST-CMD-INJECT [axes: security]
    let out = Command::new("sh")
        .arg("-c")
        .arg(user_input)
        // WAYPOINT-PLANT: RUST-UNWRAP [axes: edge-case]
        .output()
        .expect("command failed to start");
    String::from_utf8_lossy(&out.stdout).to_string()
}

// WAYPOINT-OK: fixed argv, no shell, no user-controlled interpolation
pub fn list_dir() -> std::io::Result<std::process::Output> {
    Command::new("ls").arg("-la").output()
}
